spark/src/main/scala/io/substrait/spark/expression/ToSubstraitLiteral.scala

```diff
@@ -139,15 +124,45 @@ class ToSubstraitLiteral {
         .convert(literal.dataType, nullable = true)
         .map(typedNull)
     } else {
-      convertWithValue(literal)
+      convertWithValue(literal, nullable = false)
     }
   }
 
-  def apply(literal: Literal): SExpression.Literal = {
-    convert(literal)
-      .getOrElse(
-        throw new UnsupportedOperationException(
-          s"Unable to convert the type ${literal.dataType.typeName}"))
+  /**
+   * Converts a Spark Literal to a Substrait Literal.
+   *
+   * @param literal
+   *   The Spark literal to convert
+   * @param nullable
+   *   Optional explicit nullability to use for the Substrait literal (typically from the schema).
+   *   If None, uses the literal's inferred nullability.
+   */
+  def apply(literal: Literal, nullable: Option[Boolean] = None): SExpression.Literal = {
+    nullable match {
+      case Some(n) =>
+        // Use explicit nullability
+        if (literal.value == null) {
+          if (!n) {
+            throw new IllegalArgumentException("Cannot create a non-nullable type for a null value")
+          }
+          ToSubstraitType
+            .convert(literal.dataType, nullable = true)
+            .map(typedNull)
+            .getOrElse(throw new UnsupportedOperationException(
+              s"Unable to convert the type ${literal.dataType.typeName}"))
+        } else {
+          convertWithValue(literal, n)
+            .getOrElse(
+              throw new UnsupportedOperationException(
+                s"Unable to convert the type ${literal.dataType.typeName}"))
+        }
```

Comment:
Purely a suggestion, but I wonder if the code would be clearer if the nested `if` blocks were pulled up into separate `case` conditions:
```suggestion
      case Some(false) if (literal.value == null) =>
        throw new IllegalArgumentException("Cannot create a non-nullable type for a null value")
      case Some(true) if (literal.value == null) =>
        ToSubstraitType
          .convert(literal.dataType, nullable = true)
          .map(typedNull)
          .getOrElse(throw new UnsupportedOperationException(
            s"Unable to convert the type ${literal.dataType.typeName}"))
      case Some(n) =>
        // Use explicit nullability
        convertWithValue(literal, n)
          .getOrElse(
            throw new UnsupportedOperationException(
              s"Unable to convert the type ${literal.dataType.typeName}"))
```